File: rust/links-notation/tests/empty_reference_tests.rs
//! Conformance tests for the empty reference (issue #288).
//!
//! A bare delimiter pair is the empty reference. The three delimiters `"`, `'`
//! and `` ` `` behave identically, and every longer n-quote run keeps the
//! meaning it already had. The table below is shared with the JavaScript,
//! Python, C#, Go, Java and PHP suites, so a document written by one
//! implementation reads the same in all of them.

use links_notation::{format_links, parse_lino_to_links, LiNo};

/// Render a parsed node unambiguously: every reference is wrapped in angle
/// brackets so an empty one is visible as `<>`.
fn render(node: &LiNo<String>) -> String {
    match node {
        LiNo::Ref(id) => format!("<{}>", id),
        LiNo::Link { id, values } => {
            let head = id
                .as_ref()
                .map(|id| format!("<{}>: ", id))
                .unwrap_or_default();
            let body = values.iter().map(render).collect::<Vec<_>>().join(" ");
            format!("({}{})", head, body)
        }
    }
}

fn rendered(input: &str) -> String {
    let links = parse_lino_to_links(input)
        .unwrap_or_else(|e| panic!("expected {:?} to parse, got {}", input, e));
    links.iter().map(render).collect::<Vec<_>>().join("\n")
}

fn assert_parses_as(input: &str, expected: &str) {
    assert_eq!(rendered(input), expected, "input: {}", input);
}

#[test]
fn bare_delimiter_pair_is_the_empty_reference() {
    assert_parses_as(r#"(a "" b)"#, "(<a> <> <b>)");
}

#[test]
fn every_delimiter_style_yields_the_same_empty_reference() {
    assert_parses_as(r#"(a "" b)"#, "(<a> <> <b>)");
    assert_parses_as(r#"(a '' b)"#, "(<a> <> <b>)");
    assert_parses_as("(a `` b)", "(<a> <> <b>)");
}

#[test]
fn adjacent_empty_references_stay_separate() {
    assert_parses_as(r#"(a "" "" b)"#, "(<a> <> <> <b>)");
    assert_parses_as(r#"(a '' '' b)"#, "(<a> <> <> <b>)");
    assert_parses_as("(a `` `` b)", "(<a> <> <> <b>)");
    assert_parses_as(r#"(a ""  "" b)"#, "(<a> <> <> <b>)");
}

#[test]
fn nested_empty_references_parse() {
    assert_parses_as(r#"("" ("" 1))"#, "(<> (<> <1>))");
    assert_parses_as(r#"("" ('' 1))"#, "(<> (<> <1>))");
    assert_parses_as(r#"("x" ("" 1))"#, "(<x> (<> <1>))");
    assert_parses_as(r#"("" ("x" 1))"#, "(<> (<x> <1>))");
    assert_parses_as(r#"("" x ("" 1))"#, "(<> <x> (<> <1>))");
    assert_parses_as(r#"("" 1 ("" 1))"#, "(<> <1> (<> <1>))");
}

#[test]
fn empty_reference_is_valid_as_an_id() {
    assert_parses_as(r#"("": 1)"#, "(<>: <1>)");
    assert_parses_as(r#"(o: ("" (o: ("" 1))))"#, "(<o>: (<> (<o>: (<> <1>))))");
}

#[test]
fn n_quote_delimited_bodies_are_unchanged() {
    // A run that encloses a substantive body keeps its n-quote meaning.
    assert_parses_as(r#"(a ""x"" b)"#, "(<a> <x> <b>)");
    assert_parses_as(r#"(x "" " "")"#, r#"(<x> < " >)"#);
    assert_parses_as(r#"(x ' " ')"#, r#"(<x> < " >)"#);
    // An n-quote-delimited empty is still empty.
    assert_parses_as(r#"(a """" b)"#, "(<a> <> <b>)");
}

#[test]
fn a_single_space_still_reads_as_a_space() {
    assert_parses_as(r#"(a " " b)"#, "(<a> < > <b>)");
}

#[test]
fn odd_delimiter_runs_stay_literal_text() {
    assert_parses_as(r#"(a " b)"#, r#"(<a> <"> <b>)"#);
    assert_parses_as(r#"(a """ b)"#, r#"(<a> <"""> <b>)"#);
}

#[test]
fn empty_reference_survives_a_round_trip() {
    for input in [
        r#"(a "" b)"#,
        r#"(a "" "" b)"#,
        r#"("" ("" 1))"#,
        r#"("": 1)"#,
        r#"(o: ("" (o: ("" 1))))"#,
    ] {
        let links = parse_lino_to_links(input).expect("parses");
        let formatted = format_links(&links);
        let reparsed = parse_lino_to_links(&formatted)
            .unwrap_or_else(|e| panic!("formatted {:?} did not parse: {}", formatted, e));
        assert_eq!(links, reparsed, "round trip changed {:?}", input);
    }
}

#[test]
fn empty_reference_is_written_as_a_delimiter_pair() {
    let links = parse_lino_to_links(r#"(a "" b)"#).expect("parses");
    assert_eq!(format_links(&links), r#"(a "" b)"#);
}

File: csharp/Link.Foundation.Links.Notation.Tests/EmptyReferenceTests.cs
using System.Collections.Generic;
using System.Linq;
using Xunit;

namespace Link.Foundation.Links.Notation.Tests
{
    /// <summary>
    /// Conformance tests for the empty reference.
    ///
    /// https://github.com/link-foundation/links-notation/issues/288
    ///
    /// A bare delimiter pair is the empty reference. The three delimiters
    /// <c>"</c>, <c>'</c> and <c>`</c> behave identically, and every longer
    /// n-quote run keeps the meaning it already had. The table below is shared
    /// with the Rust, JavaScript, Python, Go, Java and PHP suites, so a document
    /// written by one implementation reads the same in all of them.
    /// </summary>
    public static class EmptyReferenceTests
    {
        /// <summary>
        /// Renders a parsed node unambiguously: every reference is wrapped in
        /// angle brackets so an empty one is visible as &lt;&gt;.
        /// </summary>
        private static string Render(Link<string> node)
        {
            if (node.Values == null || node.Values.Count == 0)
            {
                return "<" + (node.Id ?? "") + ">";
            }
            var head = node.Id == null ? "" : "<" + node.Id + ">: ";
            return "(" + head + string.Join(" ", node.Values.Select(Render)) + ")";
        }

        private static string Rendered(string source)
        {
            var links = new Parser().Parse(source);
            return string.Join("\n", links.Select(Render));
        }

        private static void AssertParsesAs(string source, string expected)
        {
            Assert.Equal(expected, Rendered(source));
        }

        [Fact]
        public static void BareDelimiterPairIsTheEmptyReference()
        {
            AssertParsesAs("(a \"\" b)", "(<a> <> <b>)");
        }

        [Fact]
        public static void EveryDelimiterStyleYieldsTheSameEmptyReference()
        {
            AssertParsesAs("(a \"\" b)", "(<a> <> <b>)");
            AssertParsesAs("(a '' b)", "(<a> <> <b>)");
            AssertParsesAs("(a `` b)", "(<a> <> <b>)");
        }

        [Fact]
        public static void AdjacentEmptyReferencesStaySeparate()
        {
            AssertParsesAs("(a \"\" \"\" b)", "(<a> <> <> <b>)");
            AssertParsesAs("(a '' '' b)", "(<a> <> <> <b>)");
            AssertParsesAs("(a `` `` b)", "(<a> <> <> <b>)");
            AssertParsesAs("(a \"\"  \"\" b)", "(<a> <> <> <b>)");
        }

        [Fact]
        public static void NestedEmptyReferencesParse()
        {
            AssertParsesAs("(\"\" (\"\" 1))", "(<> (<> <1>))");
            AssertParsesAs("(\"\" ('' 1))", "(<> (<> <1>))");
            AssertParsesAs("(\"x\" (\"\" 1))", "(<x> (<> <1>))");
            AssertParsesAs("(\"\" (\"x\" 1))", "(<> (<x> <1>))");
            AssertParsesAs("(\"\" x (\"\" 1))", "(<> <x> (<> <1>))");
            AssertParsesAs("(\"\" 1 (\"\" 1))", "(<> <1> (<> <1>))");
        }

        [Fact]
        public static void EmptyReferenceIsValidAsAnId()
        {
            AssertParsesAs("(\"\": 1)", "(<>: <1>)");
            AssertParsesAs("(o: (\"\" (o: (\"\" 1))))", "(<o>: (<> (<o>: (<> <1>))))");
        }

        [Fact]
        public static void NQuoteDelimitedBodiesAreUnchanged()
        {
            // A run that encloses a substantive body keeps its n-quote meaning.
            AssertParsesAs("(a \"\"x\"\" b)", "(<a> <x> <b>)");
            AssertParsesAs("(x \"\" \" \"\")", "(<x> < \" >)");
            AssertParsesAs("(x ' \" ')", "(<x> < \" >)");
            // An n-quote-delimited empty is still empty.
            AssertParsesAs("(a \"\"\"\" b)", "(<a> <> <b>)");
        }

        [Fact]
        public static void ASingleSpaceStillReadsAsASpace()
        {
            AssertParsesAs("(a \" \" b)", "(<a> < > <b>)");
        }

        [Fact]
        public static void EmptyReferenceSurvivesARoundTrip()
        {
            var sources = new[]
            {
                "(a \"\" b)",
                "(a \"\" \"\" b)",
                "(\"\" (\"\" 1))",
                "(\"\": 1)",
                "(o: (\"\" (o: (\"\" 1))))",
            };
            foreach (var source in sources)
            {
                var formatted = ((IList<Link<string>>)new Parser().Parse(source)).Format();
                var reparsed = ((IList<Link<string>>)new Parser().Parse(formatted)).Format();
                Assert.Equal(formatted, reparsed);
            }
        }

        [Fact]
        public static void EmptyReferenceIsWrittenAsADelimiterPair()
        {
            var links = (IList<Link<string>>)new Parser().Parse("(a \"\" b)");
            Assert.Equal("(a \"\" b)", links.Format());
        }
    }
}
